SelectNow_TcpIpConnection_TransferStressTest.testLargePackets

```
java.lang.AssertionError: expected:<194583> but was:<188371>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at org.junit.Assert.assertEquals(Assert.java:631)
    at com.hazelcast.nio.tcp.TcpIpConnection_TransferStressBaseTest$1.run(TcpIpConnection_TransferStressBaseTest.java:98)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:850)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:864)
    at com.hazelcast.nio.tcp.TcpIpConnection_TransferStressBaseTest.test(TcpIpConnection_TransferStressBaseTest.java:95)
    at com.hazelcast.nio.tcp.TcpIpConnection_TransferStressBaseTest.testLargePackets(TcpIpConnection_TransferStressBaseTest.java:61)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
    at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.lang.Thread.run(Thread.java:745)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-nightly/com.hazelcast$hazelcast/739/testReport/junit/com.hazelcast.nio.tcp.nonblocking/SelectNow_TcpIpConnection_TransferStressTest/testLargePackets/

--- a/hazelcast/src/test/java/com/hazelcast/nio/tcp/TcpIpConnection_TransferStressBaseTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/nio/tcp/TcpIpConnection_TransferStressBaseTest.java
@@ -21,14 +21,14 @@ import static org.junit.Assert.assertEquals;
  */
 public abstract class TcpIpConnection_TransferStressBaseTest extends TcpIpConnection_AbstractTest {
 
-    // total running time
-    private static final long DURATION_SECONDS = TimeUnit.MINUTES.toSeconds(2);
+    // total running time for writer threads
+    private static final long WRITER_THREAD_RUNNING_TIME_IN_SECONDS = TimeUnit.MINUTES.toSeconds(2);
     // maximum number of pending packets
     private static final int maxPendingPacketCount = 10000;
     // we create the payloads up front and select randomly from them. This is the number of payloads we are creating
     private static final int payloadCount = 10000;
 
-    private final AtomicBoolean stop = new AtomicBoolean();
+    private final AtomicBoolean stop = new AtomicBoolean(false);
     private DummyPayload[] payloads;
 
     @Before
@@ -40,34 +40,38 @@ public abstract class TcpIpConnection_TransferStressBaseTest extends TcpIpConnec
     @Test
     public void testTinyPackets() throws Exception {
         makePayloads(10);
-        test();
+        testPackets();
     }
 
     @Test
     public void testSmallPackets() throws Exception {
         makePayloads(100);
-        test();
+        testPackets();
     }
 
     @Test
     public void testMediumPackets() throws Exception {
         makePayloads(1000);
-        test();
+        testPackets();
     }
 
-    @Test
+    @Test(timeout = 10 * 60 * 1000)
     public void testLargePackets() throws Exception {
         makePayloads(10000);
-        test();
+        testPackets((10 * 60 * 1000) - (WRITER_THREAD_RUNNING_TIME_IN_SECONDS * 1000));
     }
 
     @Test
     public void testSemiRealisticPackets() throws Exception {
         makeSemiRealisticPayloads();
-        test();
+        testPackets();
     }
 
-    public void test() throws Exception {
+    private void testPackets() throws Exception {
+        testPackets(ASSERT_TRUE_EVENTUALLY_TIMEOUT);
+    }
+
+    private void testPackets(long verifyTimeoutInMillis) throws Exception {
         TcpIpConnection c = connect(connManagerA, addressB);
 
         WriteThread thread1 = new WriteThread(1, c);
@@ -77,7 +81,7 @@ public abstract class TcpIpConnection_TransferStressBaseTest extends TcpIpConnec
         thread1.start();
         thread2.start();
 
-        sleepAndStop(stop, DURATION_SECONDS);
+        sleepAndStop(stop, WRITER_THREAD_RUNNING_TIME_IN_SECONDS);
 
         logger.info("Done");
 
@@ -91,14 +95,25 @@ public abstract class TcpIpConnection_TransferStressBaseTest extends TcpIpConnec
         logger.info("expected normal packets: " + expectedNormalPackets);
         logger.info("expected priority packets: " + expectedUrgentPackets);
 
-        final SocketReader XReadHandler = ((TcpIpConnection) connManagerB.getConnection(addressA)).getSocketReader();
+        final SocketWriter writer = c.getSocketWriter();
+        final SocketReader reader = ((TcpIpConnection) connManagerB.getConnection(addressA)).getSocketReader();
+        long start = System.currentTimeMillis();
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() throws Exception {
-                assertEquals(expectedNormalPackets, XReadHandler.getNormalFramesReadCounter().get());
-                assertEquals(expectedUrgentPackets, XReadHandler.getPriorityFramesReadCounter().get());
+                logger.info("writer total frames pending   : " + writer.totalFramesPending());
+                logger.info("writer last write time millis : " + writer.getLastWriteTimeMillis());
+
+                logger.info("reader total frames handled   : " + reader.getNormalFramesReadCounter().get()
+                        + reader.getPriorityFramesReadCounter().get());
+                logger.info("reader last read time millis  : " + reader.getLastReadTimeMillis());
+
+                assertEquals(expectedNormalPackets, reader.getNormalFramesReadCounter().get());
+                assertEquals(expectedUrgentPackets, reader.getPriorityFramesReadCounter().get());
             }
-        });
+        }, verifyTimeoutInMillis);
+        logger.info("Waiting for pending packets to be sent and received finished in "
+                + (System.currentTimeMillis() - start) + " milliseconds");
     }
 
     private void makePayloads(int maxSize) {

--- a/hazelcast/src/test/java/com/hazelcast/test/AbstractHazelcastClassRunner.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/AbstractHazelcastClassRunner.java
@@ -51,6 +51,8 @@ public abstract class AbstractHazelcastClassRunner extends AbstractParameterized
     protected static final int DEFAULT_TEST_TIMEOUT_IN_SECONDS = getInteger("hazelcast.test.defaultTestTimeoutInSeconds", 300);
 
     private static final ThreadLocal<String> TEST_NAME_THREAD_LOCAL = new InheritableThreadLocal<String>();
+    private static final boolean THREAD_CPU_TIME_INFO_AVAILABLE;
+    private static final boolean THREAD_CONTENTION_INFO_AVAILABLE;
 
     static {
         String logging = "hazelcast.logging.type";
@@ -72,6 +74,28 @@ public abstract class AbstractHazelcastClassRunner extends AbstractParameterized
         int g2 = rand.nextInt(255);
         int g3 = rand.nextInt(255);
         System.setProperty("hazelcast.multicast.group", "224." + g1 + "." + g2 + "." + g3);
+
+        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+
+        boolean threadCPUTimeInfoAvailable = false;
+        if (threadMXBean.isThreadCpuTimeSupported()) {
+            try  {
+                threadMXBean.setThreadCpuTimeEnabled(true);
+                threadCPUTimeInfoAvailable = true;
+            } catch (Throwable t) {
+            }
+        }
+        THREAD_CPU_TIME_INFO_AVAILABLE = threadCPUTimeInfoAvailable;
+
+        boolean threadContentionInfoAvailable = false;
+        if (threadMXBean.isThreadContentionMonitoringSupported()) {
+            try  {
+                threadMXBean.setThreadContentionMonitoringEnabled(true);
+                threadContentionInfoAvailable = true;
+            } catch (Throwable t) {
+            }
+        }
+        THREAD_CONTENTION_INFO_AVAILABLE = threadContentionInfoAvailable;
     }
 
     /**
@@ -182,7 +206,8 @@ public abstract class AbstractHazelcastClassRunner extends AbstractParameterized
         ThreadInfo[] threadInfos = threadMXBean.dumpAllThreads(true, true);
         long currentThreadId = Thread.currentThread().getId();
         for (ThreadInfo threadInfo : threadInfos) {
-            if (threadInfo.getThreadId() == currentThreadId) {
+            long threadId = threadInfo.getThreadId();
+            if (threadId == currentThreadId) {
                 continue;
             }
             dump.append('"');
@@ -193,13 +218,20 @@ public abstract class AbstractHazelcastClassRunner extends AbstractParameterized
             dump.append("\n\tjava.lang.Thread.State: ");
             dump.append(state);
             if (threadInfo.getLockName() != null) {
-                dump.append(" on lock=").append(threadInfo.getLockName());
+                dump.append(", on lock=").append(threadInfo.getLockName());
             }
             if (threadInfo.getLockOwnerName() != null) {
-                dump.append(" owned by ").append(threadInfo.getLockOwnerName());
-                dump.append(" id=").append(threadInfo.getLockOwnerId());
+                dump.append(", owned by ").append(threadInfo.getLockOwnerName());
+                dump.append(", id=").append(threadInfo.getLockOwnerId());
             }
-
+            if (THREAD_CPU_TIME_INFO_AVAILABLE) {
+                dump.append(", cpu=").append(threadMXBean.getThreadCpuTime(threadId)).append(" nsecs");
+                dump.append(", usr=").append(threadMXBean.getThreadUserTime(threadId)).append(" nsecs");
+            }
+            if (THREAD_CONTENTION_INFO_AVAILABLE) {
+                dump.append(", blocked=").append(threadInfo.getBlockedTime()).append(" msecs");
+                dump.append(", waited=").append(threadInfo.getWaitedTime()).append(" msecs");
+            }
             StackTraceElement[] stackTraceElements = threadInfo.getStackTrace();
             for (StackTraceElement stackTraceElement : stackTraceElements) {
                 dump.append("\n\t\tat ");

